### DISCUSSION TITLE
Streamline async code

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5782] [Language Ideas]
Currently, correct way of writing async code is:
```cs
async Task<int> FooAsync(CancellationToken ct)
{
  // ...
}
async Task<string> BarAsync(CancellationToken ct)
{
  int x = await FooAsync(ct);
  return x.ToString();
}
```
You can see there is quite a lot of boilerplate. Can we simplify it to something like below?
```cs
async int Foo() // method is renamed to FooAsync behind the scenes, and CancellationToken parameter is appended
{
  // ...
}
async string Bar()
{
  int x = await Foo();
  return x.ToString();
}
```
Basically, the idea is that the second snippet is actually compiled to the first snippet:
- When `async` is encountered with any return type other than `void` or `Task`-like type, switch into new mode.
- In this mode: 
- - the word `Async` is appended to the method name (if not already there), when creating IL.
- - `CancellationToken` parameter is appended (if not already there and if it's used, implicitly or explicitly)
- - Return type is changed to `Task<type>` (or `ValueTask`? not sure on up-to-date guidance on which one is best) when creating IL, where "type" is the specified return type.
- - Any `await` methods calls look for a method with the same name; if one doesn't exist, then it looks for method with a name which is the name provided plus "Async" at the end; if one exists, use that; otherwise, produce compilation error - "neither Bar nor BarAsync found".
- - Any method calls to other `Task` returning methods automatically pass in the `CancellationToken` (if the called methods expects it).

Languages which don't support this feature are not affected - they just see correctly created methods using the syntax which is already agreed on. Ideally, for `void` return (what is now achieved by returning non-generic `Task`), we can introduce a keyword which enables this mode; something like `async void Task Foo()` or whatever.

Is this a crazy idea?

**Comment by 333fred** [comment]
> When async is encountered with any return type other than void or Task, switch into new mode.

What about `ValueTask`? Or the crazy samples where `int` has an extension `GetAwaiter()` method and can be awaited?

>  the word Async is appended to the method name (if not already there), when creating IL.

This is far from a universal standard.

> CancellationToken parameter is appended (if not already there)

Not every async method needs a CancellationToken. And even if it did, how would that token even be used in the method? Sure, we can pass it down the stack, but if no one is checking the token then what good is it doing...?

**Comment by CyrusNajmabadi** [comment]
> Return type is changed to Task<type>, when creating IL, where "type" is the specified return type.

This is a non-starter as any type can be Task-like :)

**Comment by TahirAhmadov** [comment]
> What about ValueTask? Or the crazy samples where int has an extension GetAwaiter() method and can be awaited?
> ...
> This is a non-starter as any type can be Task-like :)

I updated the OP - when I said `Task`, I meant Task-like. Also, with all due respect, this is a minutia point to pick on. It doesn't affect the remaining part of my post.

> This is far from a universal standard.

It is in the BCL, and sure - some people append "Potato" for async methods - they are not obligated to use this new syntax - and are able to keep using the "Potato" suffix.

> Not every async method needs a CancellationToken. And even if it did, how would that token even be used in the method? Sure, we can pass it down the stack, but if no one is checking the token then what good is it doing...?

Actually, every async method does need a `CancellationToken`. See this: https://github.com/dotnet/runtime/issues/11369
Also, most library async methods (database, etc.) take them, so passing them down the stack is valuable.
One might say, passing an extra parameter introduces an overhead. While that's true, it's the least of our worries in the context of async methods. Nobody cares about 0.001ns of a stack push right next to 100ms database query.
Do some people ignore the cancellation token in their long-running async methods? I'm sure they do. However, that doesn't mean we need to cater to them.

**Reply by CyrusNajmabadi** [reply]
> I updated the OP - when I said Task, I meant Task-like. Also, with all due respect, this is a minutia point to pick on. It doesn't affect the remaining part of my post.

I don't think this is minutia.  Having a method like `public async X Foo()` mean that it just returns an X in one context, but might return a `Task<X>` in another is something that could really be deeply confusing.

>Do some people ignore the cancellation token in their long-running async methods? I'm sure they do. However, that doesn't mean we need to cater to them.

This is not the reason why not all async methods need a cancellation token.  The reason not all methods need a cancellation token is because not all async methods can truly be cancelled.  There are lots of cases like this.  For example, if you're in the middle of some large mutation that you cannot rollback on cancellation. Once mutation has happened, it's very common for cancellation to now be disallowed lest it leave the system in an inconsistent state.

**Reply by 333fred** [reply]
> this is a minutia point to pick on. It doesn't affect the remaining part of my post.

It really isn't though. The point we're making is that `Task` isn't special, and the language shouldn't treat it specially. This includes synthesizing it as the return type of methods.

> It is in the BCL

It isn't even close to a standard in the BCL. Their public APIs are generally postfixed, but looking through all code in the BCL that is marked `async`, they actually have more methods _not_ postfixed with `Async`.

> and sure - some people append "Potato" for async methods - they are not obligated to use this new syntax - and are able to keep using the "Potato" suffix.

The point I'm making is that this narrows the target audience. Opinionated features (of which this definitely is) need to have a very clear majority. `Async` postfixes are not a very clear majority. Consider file-scoped namespaces as an analogy here: literally 99.9+% of C# files we could find have a single namespace in them, so an opinionated feature where only one namespace is allowed is perfectly acceptable to basically every C# user. Where that feature allows flexibility is in where `using` directives are allowed: there was a majority (it's something like 60/40 outside the namespace vs inside the namespace), but that majority was not clear enough to enforce an opinion on all users of the feature.

> Actually, every async method does need a CancellationToken. See this: https://github.com/dotnet/runtime/issues/11369

Not really. If I have an async method with one purpose, and stopping in the middle of said purpose will cause the system to be an undefined state (ie, there is no way for me to gracefully cancel), then there is no point to me taking that cancellation token. Additionally, if that cancellation token isn't in scope (ie, I can't call `token.ThrowIfCancelled()`), then what good is that token doing anyone? It will never be checked, and nothing will ever be cancelled :slightly_smiling_face:.

Pulling something from the edit to the OP:

> Return type is changed to Task<type> (or ValueTask? not sure on up-to-date guidance on which one is best)

Ah, but the thing here is that there _is_ no best. Best depends entirely on whether you, as the programmer, expect that the common path will `await` or not.

**Reply by CyrusNajmabadi** [reply]
> It is in the BCL, and sure - some people append "Potato" for async methods - they are not obligated to use this new syntax - and are able to keep using the "Potato" suffix.

i guess i don't see the value there.  I would still want and expect the consumption side to always use the `Async` suffix.  That part is, itself, enormously important so i know what's going on.  For example, if i do `Task.Factory.StartNew(() => Foo())` then this looks totally normal.  But it's actually `Task.Factory.StartNew(() => FooAsync())` then this is clearly uberbroken and can be seen to be that way jsut from visual inspection of the code.

**Reply by TahirAhmadov** [reply]
> Having a method like public async X Foo() mean that it just returns an X in one context, but might return a Task<X> in another is something that could really be deeply confusing.

Yes that can happen, but is that probable to happen? I don't think so. The type of code which needs a custom Task-like type, is also the type of code which needs fine-grained control over what's going on. That X in my case is 99% of the time a regular type, like a DTO or a primitive, etc. That's the entire point of my idea - reduce the boilerplate for the bread and butter code, not the unusual strange situations.

>  you're in the middle of some large mutation that you cannot rollback on cancellation.

Sure, I can imagine something like invoking a third-party API call which is not "rollback"-able. Still, I'm not saying anybody would be mandated to check for cancellation; just that it's available. 
If it's added manually to each method, one (very reasonably) can ask the question, these methods cannot be canceled, and therefore do I need to add that parameter. However, in my proposal, the parameter is added automatically and for free. So it's a case of, we have these few methods which cannot be canceled, there is this parameter that's available, but it's not used - is that really a problem?
And actually, I updated the OP once more - if cancellation token is not used (explicitly or implicitly), it is not added to the signature, either.

**Reply by TahirAhmadov** [reply]
> It really isn't though. The point we're making is that `Task` isn't special, and the language shouldn't treat it specially. This includes synthesizing it as the return type of methods.

I understand that and corrected my OP immediately. However, you could have said something like, "you do realize that not only `Task` is task-like, right? I think you meant 'when it's not void or task-like'?" Of course I know of the existence of `ValueTask` and that C# generally relies on shapes, not types. What I'm trying to say is, the overall idea does not hinge on this detail.

> It isn't even close to a standard in the BCL. Their public APIs are generally postfixed, but looking through all code in the BCL that is marked `async`, they actually have more methods _not_ postfixed with `Async`.

I don't know enough about their development process, maybe it can be argued that they should enforce a standard even for their internal code? Either way, they are not mandated to use this new syntax and their existing code continues building and working just like before.

> The point I'm making is that this narrows the target audience. Opinionated features (of which this definitely is) need to have a very clear majority. `Async` postfixes are not a very clear majority. 

Even if "Async" suffix is not a super-majority, it's still the only suffix in use - when one is used at all. I understand some people prefer to skip it - guess why? Because they are tired of typing it every time. My proposal addresses that - C# does it for them.

> Not really. If I have an async method with one purpose, and stopping in the middle of said purpose will cause the system to be an undefined state (ie, there is no way for me to gracefully cancel), then there is no point to me taking that cancellation token. 

I responded to Cyrus above and also modified my post - basically, the token can be added only if it's used (implicitly or explicitly). Bingo.

> Additionally, if that cancellation token isn't in scope (ie, I can't call `token.ThrowIfCancelled()`), then what good is that token doing anyone? It will never be checked, and nothing will ever be cancelled 🙂.

Can you explain? How can it not be in scope? Why can't you call `token.ThrowIfCancelled()` anywhere inside a method, if a parameter is available?

> > Return type is changed to Task (or ValueTask? not sure on up-to-date guidance on which one is best)
> 
> Ah, but the thing here is that there _is_ no best. Best depends entirely on whether you, as the programmer, expect that the common path will `await` or not.

That's a perfectly valid point. On the output side, we can either:
- Always stick with one return type; settle on `Task` or `ValueTask` and be done with it.
- Try to determine the best one by looking at the method body. For example, if more than half of downstream calls return `ValueTask`, use that, otherwise use `Task`.

**Reply by TahirAhmadov** [reply]
> i guess i don't see the value there. I would still want and expect the consumption side to always use the `Async` suffix. That part is, itself, enormously important so i know what's going on. For example, if i do `Task.Factory.StartNew(() => Foo())` then this looks totally normal. But it's actually `Task.Factory.StartNew(() => FooAsync())` then this is clearly uberbroken and can be seen to be that way jsut from visual inspection of the code.

In your example, that lambda is not in the "mode", because it doesn't have `async` keyword before it.

**Comment by HaloFour** [comment]
The design of iterators and async methods has been intentionally such that the signature that you write is the signature that you get.  Everything else is an implementation detail within the method body.  This would be a pretty big departure from that and one that I wouldn't welcome.

**Reply by HaloFour** [reply]
Also, it's perfectly legal today for an `async` method to return a `Task<Task<T>>` but that would be very ambiguous in this case.

**Reply by TahirAhmadov** [reply]
None of that changes, BTW. This is purely syntactic sugar. Like I said, nothing changes in the IL - the two snippets in my OP compile to the same thing (under my proposal). We are **not** introducing a new way to return an `int` in CLR which is magically awaitable.

Think of it as a combination of something similar to dropping the word Attribute from attribute type names; and extension method semantics WRT treating some parameters in a special way.

**Reply by TahirAhmadov** [reply]
> Also, it's perfectly legal today for an `async` method to return a `Task<Task<T>>` but that would be very ambiguous in this case.

Not really; this new "mode" is only entered if the return type is not Task-like. Obviously `Task<Task<T>>` is Task-like, so it stays in the regular, old-school mode.

**Reply by HaloFour** [reply]
That's my point, there is no mode:

```cs
public async Task<int> Foo() { ... }
```

could be either of:

```cs
public async Task<int> Foo() { ... }
```

or

```cs
public async Task<Task<int>> FooAsync(CancellationToken token) { ... }
```

and there's no way the compiler could tell which is what you actually intended.

IMO, all of the confusion disappears when the signature you write is the signature you get.  No implicit transformations.  It's a virtue.

**Reply by TahirAhmadov** [reply]
To clarify, `public Task<int> Foo() { ... }` always stays `public Task<int> Foo() { ... }` for two reasons: 1, it doesn't have `async`; 2, return type is already task-like, and therefore this new mode is not used at all.

**Reply by HaloFour** [reply]
> it doesn't have `async`

That's my fault, I forgot to put `async` there.

Either way, there is no new "mode".  You're trying to repurpose what is already perfectly legal syntax to mean something completely different.  And you're trying to silently change the signature of a method.  It's a non-starter, it completely goes against many of the tenets of the design being `async` in the first place.

**Reply by HaloFour** [reply]
But note that the removal of `async` today doesn't in any way change the signature of the method.  That is intentional.  `async` is an implementation detail.

**Reply by TahirAhmadov** [reply]
> Either way, there is no new "mode". You're trying to repurpose what is already perfectly legal syntax to mean something completely different. And you're trying to silently change the signature of a method. It's a non-starter, it completely goes against many of the tenets of the design being async in the first place.

I think you misunderstood a couple of points I made. The `async int Foo()` syntax is currently illegal (almost always - provided there is no `GetAwaiter()` extension on `int`). And regarding "silently changing" - yes, that is a partially valid criticism; however, I think if this feature is added, people will get used to it and think back to the days when they had to write out the `Task<>` and the `Async` and the `CancellationToken` as the dark ages of async :)

> But note that the removal of async today doesn't in any way change the signature of the method. That is intentional. async is an implementation detail.

Yes - in the sense that `async Task<int>` and `Task<int>` is the same thing signature wise. My idea is to allow `async int` which becomes a **signature** and an implementation detail in one. Like shampoo and conditioner :)

**Reply by HaloFour** [reply]
@TahirAhmadov 

> is currently illegal 

Literally any type is potentially a task-like.  The C# team did quite a bit of work to ensure that.

> which becomes a **signature** and an implementation detail in one.

It's not possible to be both a part of the signature and entirely independent of the signature.

> I think if this feature is added

Don't worry, it won't be.

**Reply by TahirAhmadov** [reply]
> Literally any type is potentially a task-like. The C# team did quite a bit of work to ensure that.

It's "potentially" a task-like, and the compiler knows whether it's task-like when it sees `async X`. Currently, if `X` is not task-like, it becomes a compilation error. My point is we modify that outcome to mean something new.

> It's not possible to be both a part of the signature and entirely independent of the signature.

It is not possible currently but it will be possible if my idea is implemented :)

> Don't worry, it won't be.

Trust me, I'm not holding my breathe :) But I think it was worth sharing my idea anyway.

However, I have to note that other than the obvious downside of the initial learning curve for developers, I haven't seen any good points which invalidate the idea as a whole. With that, I'm off. All the best :)

**Comment by theunrepentantgeek** [comment]
I'm pretty sure code like this was considered - and rejected - as a part of the _original design process_ for **async** and **await**.

**Reply by HaloFour** [reply]
I know I've seen it in one of the blogs on the async/await design but I haven't been able to find it.
